File: integrations/hermes/open_android_intelligence_gateway/account_paths.py
"""Portable, file-level account isolation paths for the Hermes Gateway."""

from __future__ import annotations

import hashlib
import re
from dataclasses import dataclass
from pathlib import Path


WIRE_ID_PATTERN = re.compile(r"^[A-Za-z0-9._~-]{1,128}$")


@dataclass(frozen=True)
class AccountPaths:
    root: Path
    database: Path
    attachments: Path
    audit: Path


def assert_opaque_id(account_id: str) -> None:
    if not isinstance(account_id, str) or WIRE_ID_PATTERN.fullmatch(account_id) is None:
        raise ValueError("SCHEMA_INVALID")


def sha256_hex(value: str) -> str:
    return hashlib.sha256(value.encode("utf-8")).hexdigest()


def resolve_within(root: str | Path, child: str | Path) -> Path:
    resolved_root = Path(root).resolve()
    resolved_child = (resolved_root / child).resolve()
    if resolved_child == resolved_root or resolved_root not in resolved_child.parents:
        raise ValueError("SCHEMA_INVALID")
    return resolved_child


def account_paths(storage_root: str | Path, account_id: str) -> AccountPaths:
    assert_opaque_id(account_id)
    account_root = resolve_within(storage_root, sha256_hex(account_id))
    return AccountPaths(
        root=account_root,
        database=account_root / "gateway.sqlite",
        attachments=account_root / "attachments",
        audit=account_root / "audit",
    )


def ensure_account_directories(paths: AccountPaths) -> None:
    paths.root.mkdir(parents=True, exist_ok=True, mode=0o700)
    paths.attachments.mkdir(parents=True, exist_ok=True, mode=0o700)
    paths.audit.mkdir(parents=True, exist_ok=True, mode=0o700)
    for directory in (paths.root, paths.attachments, paths.audit):
        try:
            directory.chmod(0o700)
        except OSError:
            pass


def default_hermes_gateway_root() -> Path:
    return (Path.cwd() / ".open-android-intelligence-hermes" / "accounts").resolve()


accountPaths = account_paths
assertOpaqueId = assert_opaque_id
ensureAccountDirectories = ensure_account_directories

File: integrations/openclaw/src/core/account-paths.ts
import { createHash } from "node:crypto";
import { mkdirSync } from "node:fs";
import { join, resolve, relative, sep } from "node:path";

const wireIdPattern = /^[A-Za-z0-9._~-]{1,128}$/;

export type AccountPaths = Readonly<{
  root: string;
  database: string;
  attachments: string;
  audit: string;
}>;

export const assertOpaqueId = (accountId: string): void => {
  if (!wireIdPattern.test(accountId)) throw new Error("SCHEMA_INVALID");
};

export const sha256Hex = (value: string): string =>
  createHash("sha256").update(value, "utf8").digest("hex");

export const resolveWithin = (root: string, child: string): string => {
  const resolvedRoot = resolve(root);
  const resolvedChild = resolve(resolvedRoot, child);
  const pathToChild = relative(resolvedRoot, resolvedChild);
  if (
    pathToChild.length === 0 ||
    pathToChild.startsWith("..") ||
    pathToChild.includes(`..${sep}`) ||
    resolve(pathToChild) === pathToChild
  ) {
    throw new Error("SCHEMA_INVALID");
  }
  return resolvedChild;
};

export const defaultOpenClawGatewayRoot = (): string =>
  resolve(process.cwd(), ".open-android-intelligence-openclaw", "accounts");

export const accountPaths = (root: string, accountId: string): AccountPaths => {
  assertOpaqueId(accountId);
  const accountRoot = resolveWithin(root, sha256Hex(accountId));
  return Object.freeze({
    root: accountRoot,
    database: join(accountRoot, "gateway.sqlite"),
    attachments: join(accountRoot, "attachments"),
    audit: join(accountRoot, "audit"),
  });
};

export const ensureAccountDirectories = (paths: AccountPaths): void => {
  mkdirSync(paths.root, { recursive: true, mode: 0o700 });
  mkdirSync(paths.attachments, { recursive: true, mode: 0o700 });
  mkdirSync(paths.audit, { recursive: true, mode: 0o700 });
};
